lang/py/avro/compatibility.py

```diff
@@ -0,0 +1,314 @@
+                        result = result.merged_with(
+                            incompatible(
+                                SchemaIncompatibilityType.missing_union_branch,
+                                f"reader union lacking writer type: {writer_branch.type.upper()}", location + [f"{i}"]
+                            )
+                        )
+                return result
+            raise AvroRuntimeException(f"Unknown schema type: {reader.type}")
+        if writer.type == SchemaType.UNION:
+            writer = cast(UnionSchema, writer)
+            for s in writer.schemas:
+                result = result.merged_with(self.get_compatibility(reader, s))
+            return result
+        if reader.type in {SchemaType.NULL, SchemaType.BOOLEAN, SchemaType.INT}:
+            return result.merged_with(self.type_mismatch(reader, writer, location))
+        if reader.type == SchemaType.LONG:
+            if writer.type == SchemaType.INT:
+                return result
+            return result.merged_with(self.type_mismatch(reader, writer, location))
+        if reader.type == SchemaType.FLOAT:
+            if writer.type in {SchemaType.INT, SchemaType.LONG}:
+                return result
+            return result.merged_with(self.type_mismatch(reader, writer, location))
+        if reader.type == SchemaType.DOUBLE:
+            if writer.type in {SchemaType.INT, SchemaType.LONG, SchemaType.FLOAT}:
+                return result
+            return result.merged_with(self.type_mismatch(reader, writer, location))
+        if reader.type == SchemaType.BYTES:
+            if writer.type == SchemaType.STRING:
+                return result
+            return result.merged_with(self.type_mismatch(reader, writer, location))
+        if reader.type == SchemaType.STRING:
+            if writer.type == SchemaType.BYTES:
+                return result
+            return result.merged_with(self.type_mismatch(reader, writer, location))
+        if reader.type in {SchemaType.ARRAY, SchemaType.MAP, SchemaType.FIXED, SchemaType.ENUM, SchemaType.RECORD}:
+            return result.merged_with(self.type_mismatch(reader, writer, location))
+        if reader.type == SchemaType.UNION:
+            reader = cast(UnionSchema, reader)
+            for reader_branch in reader.schemas:
+                compat = self.get_compatibility(reader_branch, writer)
+                if compat.compatibility is SchemaCompatibilityType.compatible:
+                    return result
+            # No branch in reader compatible with writer
+            message = f"reader union lacking writer type {writer.type}"
+            return result.merged_with(
+                incompatible(
+                    SchemaIncompatibilityType.missing_union_branch, message, location
+                )
+            )
+        raise AvroRuntimeException(f"Unknown schema type: {reader.type}")
+
+    # pylSchemaType.INT: enable=too-many-return-statements
+
+    @staticmethod
```

Comment:
There still seem to be some staticmethods in here :/

Reply:
sorry... missed those ones...  fixed now